### PR TITLE
make sure socket is blocking after accept() on macOS and BSD

### PR DESCRIPTION
**Pull request by wfurt**
This change will make sure new sockets after Accept() are in blocking mode on macOS. 
See the notes from #27210. 

It will also make them auto-close after exec to match Linux behavior.
I have no direct evidence that it is causing problems but now remote exec xUnit tests will inherit sockets from main process.

### CODE REVIEW

**Comment by wfurt** [comment]
@dotnet-bot test outerloop Linux x64 debug build
@dotnet-bot test outerloop Windows x64 debug build
@dotnet-bot test outerloop OSX x64 debug build

File: src/Native/Unix/System.Native/pal_networking.c [hunk 1]
```diff
@@ -1650,12 +1651,35 @@ int32_t SystemNative_Accept(intptr_t socket, uint8_t* socketAddress, int32_t* so
...
+#if defined(FD_CLOEXEC)
+    // macOS does not have accept4 but it can set _CLOEXEC on descriptor.
+    if (accepted != -1)
+    {
+        if (fcntl(accepted, F_SETFD, FD_CLOEXEC) != 0)
+        {
+            int oldError = errno;
+            close(accepted);
+            *acceptedSocket = -1;
+            return SystemNative_ConvertErrorPlatformToPal(oldError);
```
Comment (stephentoub) [inline src/Native/Unix/System.Native/pal_networking.c]: We could delete this line and just let it fall through.

File: src/Native/Unix/System.Native/pal_networking.c [hunk 2]
```diff
@@ -1650,12 +1651,35 @@ int32_t SystemNative_Accept(intptr_t socket, uint8_t* socketAddress, int32_t* so
...
+#if !defined(__linux__)
+    // On macOS and FreeBSD new socket inherits flags from accepting fd.
+    if (SystemNative_FcntlSetIsNonBlocking(accepted, 0) != 0)
+    {
+        int oldError = errno;
+        close(accepted);
+        *acceptedSocket = -1;
+        return SystemNative_ConvertErrorPlatformToPal(oldError);
+    }
+#endif
```
Comment (stephentoub) [inline src/Native/Unix/System.Native/pal_networking.c]: Similarly we could move this block up to above the -1 check above, delete this return line, and just have all of the -1 checking and error code handling consolidated.

**Comment by geoffkizer** [comment]
Thanks for tracking this down!  Aside from Stephen's feedback, LGTM.

**Comment by tmds** [comment]
Cause and fix LGTM. This is indeed due to BSD inheriting the listen socket flags, while this is not the case on Linux.

File: src/Native/Unix/System.Native/pal_networking.c [hunk 3]
```diff
@@ -1650,12 +1651,35 @@ int32_t SystemNative_Accept(intptr_t socket, uint8_t* socketAddress, int32_t* so
     while ((accepted = accept4(fd, (struct sockaddr*)socketAddress, &addrLen, SOCK_CLOEXEC)) < 0 && errno == EINTR);
 #else
     while ((accepted = accept(fd, (struct sockaddr*)socketAddress, &addrLen)) < 0 && errno == EINTR);
+#if defined(FD_CLOEXEC)
+    // macOS does not have accept4 but it can set _CLOEXEC on descriptor.
```
Comment (tmds) [inline src/Native/Unix/System.Native/pal_networking.c]: Maybe add a note that this is _best effort_. An exec between accept and fcntl will still inherit the socket.

Comment (wfurt) [reply src/Native/Unix/System.Native/pal_networking.c]: good point since the calls are not atomic unlike accept4. Should we than simply ignore failures?

Comment (tmds) [reply src/Native/Unix/System.Native/pal_networking.c]: Or `assert` the return value is `0` since the call should never fail.

**Comment by wfurt** [comment]
@dotnet-bot test outerloop Linux x64 debug build
@dotnet-bot test outerloop Windows x64 debug build
@dotnet-bot test outerloop OSX x64 debug build

**Comment by wfurt** [comment]
I  changed the code. Just a note that both new blocks are under different #if as well as I think we should preserve errno from failed system call instead of close() That made fall-through little bit more complicated. 

I was also thinking about note from @tmds.  Since new code is not atomic as accept4, we could possibly modify flags on accepting socket to inherit FD_CLOEXEC. AT this point it would be only problem is macOS as FreeBSD has accept4 with similar functionality as Linux.

File: src/Native/Unix/System.Native/pal_networking.c [hunk 4]
```diff
@@ -1255,6 +1256,29 @@ int32_t SystemNative_Accept(intptr_t socket, uint8_t* socketAddress, int32_t* so
...
+    // Unlike accept4 it is not atomic and the fd can leak child process.
```
Comment (stephentoub) [inline src/Native/Unix/System.Native/pal_networking.c]: Nit: "the fd can leak child process" => "the fd can leak into the child process"

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs [hunk 5]
```diff
@@ -1013,10 +1013,7 @@ public SocketAsyncContext(SafeCloseSocket socket)
 
         private void Register()
         {
-            // Note, on OSX, this is not always true because in certain cases, 
-            // the socket can already be in non-blocking mode even though we didn't set that ourselves.
-            // TODO: Track down exactly why this is
-            // Debug.Assert(_nonBlockingSet);
+            Debug.Assert(_nonBlockingSet);
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs]: I understand the changes made in the native code, but how does that affect the value of _nonBlockingSet in the managed code?

Comment (geoffkizer) [reply src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs]: This code gets executed when we get EAGAIN from a socket operation like send() or recv().

When this happens, we expect we've previously set the socket to non-blocking.  That's what the assert is testing.

Comment (stephentoub) [reply src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs]: Ok.  But if this assert was previously failing, that means we were hitting this code path due to such an EAGAIN from send/recv.  With the changes made in the .c file, we will still hit this code path for an EAGAIN, right?  What in the .c change will either case _nonBlockingSet to be true now when it previously would have been false, or will change us not to follow this code path when it is false?  I'm obviously missing something obvious :)

Comment (geoffkizer) [reply src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs]: The changes in the .c file should ensure the socket is not in non-blocking mode unless we explicitly set it to non-blocking mode (at which time we will set _nonBlockingSet).

So we should not receive EAGAIN unless we've explicitly set the socket to non-blocking mode.

Comment (stephentoub) [reply src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs]: Aha.  Got it.  So it's the "will change us not to follow this code path when it is false" case.  Thanks.

**Comment by wfurt** [comment]
OSX is failing in FileSystem tests. Seems unrelated to this PR.

**Pull request merged**